File: codes/3.ProgrammingLanguage/编程语言试水/Score100WordsAnalyzer.js
const fs = require("fs") // node 写法，性能稍差，运行：node Score100WordsAnalyzer.js
// import fs from 'node:fs' // deno 写法，性能更好，运行：deno run --allow-read=. Score100WordsAnalyzer.js

const start = Date.now();
const score100Words = fs
  .readFileSync("Dictionary.txt", "utf8")
  .split("\r\n")
  .filter(
    (line) => 
      !line.includes("-") 
      && !line.includes(" ") 
      && !line.includes("="),
  )
  .filter(
    (word) =>
      word
        .split("")
        .reduce(
          (sum, c) => 
            sum + c.charCodeAt() - "a".charCodeAt() + 1, 
          0) === 100
  );
console.log(score100Words.join(","));
console.log(`\nwords number: ${score100Words.length}`);
console.log("        time: " + (Date.now() - start) + "ms");

/*
编码体验：
1、句尾的分号很随意，可以有，也可以没有
2、有变量类型
3、校验很松散
*/

File: codes/3.ProgrammingLanguage/编程语言试水/Score100WordsAnalyzer.clj
(def start (System/currentTimeMillis))
(def lines (clojure.string/split-lines (slurp "Dictionary.txt")))
(defn calculate-score [line] (reduce + (map #(+ (- (int %) (int \a)) 1) line)))
(def score-100-words (filter #(and (not (clojure.string/includes? % " "))
                                   (not (clojure.string/includes? % "-"))
                                   (not (clojure.string/includes? % "="))
                                   (= (calculate-score %) 100)) lines))
(println score-100-words)
(println "\nwords number: " (count score-100-words))
(println "        time: " (- (System/currentTimeMillis) start) "ms")

; 编码体验
; 加分项：
; 1、没有分号
; 2、可以直接调用Java类库
; 3、读取文本文件很方便
; 4、字符串连接很方便
; 5、代码很简洁，跟Java比少了很多代码
;
; 减分项：
; 1、需要很多圆括号
; 2、调用方法时前面又有点又有斜杠，如：clojure.string/split-lines
; 3、运行符需要放在前面，如：(= (calculate-score %) 100)
;
